### DISCUSSION TITLE
WS 1006: Report OS error?

### DISCUSSION THREAD

**Discussion by FGasper** [mojolicious/mojo#1754] [General]
Hi.

I have a WS client that’s reporting 1006 as its failure code. I strace’d and saw that it’s because the socket is getting ECONNRESET, but that’s not being reported as the reason for the 1006 error code.

Is there any interface that Mojo exposes that would tell me why a WS 1006 status happened? I could read `$!`, but that’s assuming that no other system call happened between whatever failure and the Perl code … which, I don’t know how safe that is.

Thank you!

(P.S.: Does anyone know a reason why _both_ ends of a TCP connection would report ECONNRESET?)

**Comment by christopherraa** [comment]
For what it is worth: we keep seeing this. Every time we've done any extensive debugging it has been misbehaving "security" products on the clients end that just drops the traffic both ways. Someone else would have to answer if Mojolicious could detect this, but I suspect the answer is no.

**Reply by FGasper** [reply]
> we keep seeing this

The double-sided ECONNRESET? Good to know!

Mojo can’t detect the double-reset, but it could at least somehow expose the ECONNRESET that its own I/O operation gets. Right now all I’m seeing is a 1006.

**Reply by kraih** [reply]
Suggestions for improvements would be very welcome.

**Reply by FGasper** [reply]
@kraih https://github.com/mojolicious/mojo/blob/103b2214cd176c2987ff75e4e8e05492509e3403/lib/Mojo/IOLoop/Stream.pm#L104-L117

^^ It actually looks like ECONNRESET is purposely “thrown away” … maybe just allow ECONNRESET to go out in an `error` event the same as other such errors?

**Reply by FGasper** [reply]
Looks like this behaviour of ignoring ECONNRESET dates from pre-1.0: https://github.com/mojolicious/mojo/commit/a21fe0a4ee8667e936bbd215518f03b4e13a6315

It makes sense for a server, but for a client doesn’t it make more sense to report abnormal connection termination than to ignore it?
